examples/hotrod/pkg/pool/pool_test.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2020 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pool
+
+import (
+	"sync"
+	"sync/atomic"
+	"testing"
+)
+
+var n int32 = 0
+var wg sync.WaitGroup
+
+func testJob() {
+	atomic.AddInt32(&n, 1)
+	wg.Done()
+}
+
+func TestExecute(t *testing.T) {
+	tests := []struct {
+		name     string
+		pool     *Pool
+		job      func()
+		expected int32
+	}{
+		{
+			name:     "worker: 1, job: 1",
+			pool:     New(1),
+			job:      testJob,
+			expected: 1,
+		},
+		{
+			name:     "worker: 2, job: 2",
+			pool:     New(2),
+			job:      testJob,
+			expected: 2,
+		},
+		{
+			name:     "worker: 3, job: 3",
+			pool:     New(3),
+			job:      testJob,
+			expected: 3,
+		},
+		{
+			name:     "worker: 4, job: 4",
+			pool:     New(4),
+			job:      testJob,
+			expected: 4,
+		},
+		{
+			name:     "worker: 1, job: 5",
+			pool:     New(1),
+			job:      testJob,
+			expected: 5,
+		},
+		{
+			name:     "worker: 2, job: 6",
+			pool:     New(2),
+			job:      testJob,
+			expected: 6,
+		},
+	}
+	for index, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			wg.Add(index + 1)
+			n = 0
+			for i := 0; index >= i; i++ {
+				tt.pool.Execute(tt.job)
+			}
+			wg.Wait()
+			tt.pool.Stop()
+			if n != tt.expected{
```

Comment:
Note that this is not really testing the concurrent of the executor pool. For example, if you have a pool with 2 goroutines, and you run two jobs, this test cannot verify that both jobs did not run sequentially in a single goroutine rather than being run concurrently.

Reply:
I try to use a  rwlock instead of sleep, but when the number of jobs is greater than the number of workers, the program will be blocked.